yoyodyne/evaluators.py

```diff
@@ -155,3 +213,136 @@ def finalize_predictions(
+    def name(self) -> str:
+        return "accuracy"
+
+
+class SEREvaluator(Evaluator):
+    """Evaluates symbol error rate.
+    
+    Here, a symbol is defined by the user specified tokenization."""
+
+    def _compute_ser(
+        self,
+        preds: List[str],
+        target: List[str],
+    ) -> float:
+        errors = _edit_distance(preds, target)
+        total = len(target)
+        return errors / total
+
+    def get_eval_item(
+        self,
+        predictions: List[List[str]],
+        golds: List[List[str]],
+        pad_idx: int,
+    ) -> EvalItem:
+        sers = [self._compute_ser(p, g) for p, g in zip(predictions, golds)]
+        return EvalItem(sers)
+
+    def _finalize_tensor(
+        self,
+        tensor: torch.Tensor,
+        end_idx: int,
+        pad_idx: int,
+    ) -> List[List[str]]:
+        # Not necessary if batch size is 1.
+        if tensor.size(0) == 1:
+            return [numpy.char.mod("%d", tensor.cpu().numpy())]
+        out = []
+        for i, prediction in enumerate(tensor):
+            # Gets first instance of EOS.
+            eos = (prediction == end_idx).nonzero(as_tuple=False)
+            if len(eos) > 0 and eos[0].item() < len(prediction):
+                # If an EOS was decoded and it is not the last one in the
+                # sequence.
+                eos = eos[0]
+            else:
+                # Leaves tensor[i] alone.
+                out.append(numpy.char.mod("%d", prediction))
+                continue
+            # Hack in case the first prediction is EOS. In this case
+            # torch.split will result in an error, so we change these 0's to
+            # 1's, which will make the entire sequence EOS as intended.
+            eos[eos == 0] = 1
+            symbols, *_ = torch.split(prediction, eos)
+            out.append(numpy.char.mod("%d", symbols))
+        return out
+
+    def finalize_predictions(
+        self,
+        predictions: torch.Tensor,
+        end_idx: int,
+        pad_idx: int,
+    ) -> List[List[str]]:
+        """Finalizes predictions.
+
+        Args:
+            predictions (torch.Tensor): prediction tensor.
+            end_idx (int).
+            pad_idx (int).
+
+        Returns:
+            torch.Tensor: finalized predictions.
+        """
+        return self._finalize_tensor(predictions, end_idx, pad_idx)
+
+    def finalize_golds(
+        self,
+        golds: torch.Tensor,
+        end_idx: int,
+        pad_idx: int,
+    ):
+        return self._finalize_tensor(golds, end_idx, pad_idx)
+
+    @property
+    def name(self) -> str:
+        return "ser"
+
+
+_eval_factory = {
+    "accuracy": AccuracyEvaluator,
+    "ser": SEREvaluator,
+}
+
+
+def get_evaluator(eval_metric: str) -> Evaluator:
+    """Gets the requested Evaluator given the specified metric.
+
+    Args:
+        eval_metric (str).
+
+    Raises:
+        Error.
+
+    Returns:
+        Evaluator.
+    """
+    if eval_metric not in _eval_factory:
```

Comment:
The slightly more Pythonic, and faster, thing to do here is:

```python
try:
    return _eval_factory[eval_metric]
except KeyError(key):
    raise Error(f"No evaluation metric {key}")
```

This is thought better because you're not doing the lookup twice (even if it is constant-time...)

Reply:
Yeah I usually do that, for some reason I was thinking I had been told somewhere that the if statement was considered clearer or cleaner, but you are right, I will change this!